File: backend/app/services/graph_builder.py
"""
Graph builder service for constructing temporal communication graphs.

Takes parsed messages and builds a NetworkX graph with:
- Nodes representing agents
- Edges with temporal interaction history
- Support for time-based filtering
"""

from datetime import datetime
from typing import Dict, List, Optional, Set, Tuple
import networkx as nx
from app.models.types import (
    EdgeData,
    GraphSnapshot,
    Interaction,
    IntentLabel,
    Message,
    NodeData,
)


class GraphBuilder:
    """
    Service for building temporal communication graphs from message lists.

    The graph uses a time-aware edge model where edges store a list of
    Interaction objects, enabling time-slider visualization and temporal analysis.
    """

    def __init__(self):
        self.graph: Optional[nx.DiGraph] = None
        self.messages: List[Message] = []

    def build_graph(self, messages: List[Message]) -> nx.DiGraph:
        """
        Build a directed graph from a list of messages.

        Args:
            messages: List of Message objects in chronological order

        Returns:
            NetworkX DiGraph with temporal edge data
        """
        if not messages:
            raise ValueError("Cannot build graph from empty message list")

        self.messages = sorted(messages, key=lambda m: m.step_index)
        self.graph = nx.DiGraph()

        # Track node metadata
        node_metadata: Dict[str, Dict] = {}

        # Build nodes and edges
        for message in self.messages:
            # Add sender node
            self._add_or_update_node(message.sender, message, node_metadata)

            # Add edge if there's a receiver
            if message.receiver:
                # Add receiver node
                self._add_or_update_node(message.receiver, message, node_metadata)

                # Add interaction to edge
                self._add_interaction(message)

        # Convert node metadata to NodeData and attach to graph
        for node_id, metadata in node_metadata.items():
            self.graph.nodes[node_id].update(metadata)

        return self.graph

    def _add_or_update_node(
        self,
        node_id: str,
        message: Message,
        node_metadata: Dict[str, Dict]
    ) -> None:
        """
        Add a new node or update existing node metadata.

        Args:
            node_id: The agent/node identifier
            message: The message being processed
            node_metadata: Dictionary tracking node metadata
        """
        if node_id not in node_metadata:
            node_metadata[node_id] = {
                'id': node_id,
                'label': node_id,
                'message_count': 0,
                'first_appearance': message.timestamp,
                'last_activity': message.timestamp,
                'metadata': {}
            }
            self.graph.add_node(node_id)

        # Update metadata
        if message.sender == node_id:
            node_metadata[node_id]['message_count'] += 1
            node_metadata[node_id]['last_activity'] = max(
                node_metadata[node_id]['last_activity'],
                message.timestamp
            )

    def _add_interaction(self, message: Message) -> None:
        """
        Add an interaction to the edge between sender and receiver.

        Args:
            message: The message creating this interaction
        """
        sender = message.sender
        receiver = message.receiver

        if not receiver:
            return

        # Create interaction object
        interaction = Interaction(
            step_index=message.step_index,
            timestamp=message.timestamp,
            intent=self._infer_intent(message),
            message_id=message.step_index,
            metadata={
                'message_type': message.message_type.value,
                'content_preview': message.content[:100] if message.content else ''
            }
        )

        # Add or update edge
        if self.graph.has_edge(sender, receiver):
            # Edge exists - append interaction
            self.graph[sender][receiver]['interactions'].append(interaction)
            self.graph[sender][receiver]['weight'] += 1
        else:
            # New edge
            self.graph.add_edge(
                sender,
                receiver,
                interactions=[interaction],
                weight=1,
                source=sender,
                target=receiver
            )

    def _infer_intent(self, message: Message) -> IntentLabel:
        """
        Infer the intent label from a message.

        This is a simple rule-based approach. In Phase 3, this will be
        replaced with LLM-based abstraction.

        Args:
            message: The message to analyze

        Returns:
            IntentLabel enum value
        """
        message_type = message.message_type

        # Map message types to intent labels
        type_to_intent = {
            'delegation': IntentLabel.DELEGATION,
            'action': IntentLabel.COORDINATION,
            'response': IntentLabel.INFORMATION_RESPONSE,
            'feedback': IntentLabel.FEEDBACK,
        }

        return type_to_intent.get(message_type.value, IntentLabel.UNKNOWN)

    def filter_graph_by_step(self, max_step: int) -> nx.DiGraph:
        """
        Return a filtered graph containing only interactions up to max_step.

        This is critical for time-slider functionality.

        Args:
            max_step: Maximum step index to include

        Returns:
            New NetworkX DiGraph filtered to the specified step
        """
        if not self.graph:
            raise ValueError("No graph built yet. Call build_graph() first.")

        filtered = self.graph.copy()

        # Filter edges
        edges_to_remove = []
        for u, v, data in filtered.edges(data=True):
            interactions = [
                i for i in data['interactions']
                if i.step_index <= max_step
            ]

            if interactions:
                filtered[u][v]['interactions'] = interactions
                filtered[u][v]['weight'] = len(interactions)
            else:
                edges_to_remove.append((u, v))

        # Remove edges with no interactions
        for u, v in edges_to_remove:
            filtered.remove_edge(u, v)

        # Remove isolated nodes (nodes with no edges)
        isolated_nodes = list(nx.isolates(filtered))
        filtered.remove_nodes_from(isolated_nodes)

        return filtered

    def to_graph_snapshot(self, max_step: Optional[int] = None) -> GraphSnapshot:
        """
        Convert the graph to a GraphSnapshot for API response.

        Args:
            max_step: Optional maximum step to filter to

        Returns:
            GraphSnapshot object ready for JSON serialization
        """
        if not self.graph:
            raise ValueError("No graph built yet. Call build_graph() first.")

        # Use filtered graph if max_step provided
        graph_to_use = self.filter_graph_by_step(max_step) if max_step else self.graph

        # Convert nodes
        nodes = []
        for node_id in graph_to_use.nodes():
            node_data_dict = graph_to_use.nodes[node_id]
            nodes.append(NodeData(**node_data_dict))

        # Convert edges
        edges = []
        for u, v, data in graph_to_use.edges(data=True):
            edges.append(EdgeData(
                source=data['source'],
                target=data['target'],
                interactions=data['interactions'],
                weight=data['weight']
            ))

        total_steps = max(m.step_index for m in self.messages) if self.messages else 0

        return GraphSnapshot(
            nodes=nodes,
            edges=edges,
            current_step=max_step,
            total_steps=total_steps,
            metadata={
                'node_count': len(nodes),
                'edge_count': len(edges),
                'message_count': len(self.messages)
            }
        )

    def get_graph_metrics(self) -> Dict:
        """
        Calculate basic graph metrics.

        Returns:
            Dictionary of metrics (centrality, density, etc.)
        """
        if not self.graph:
            raise ValueError("No graph built yet. Call build_graph() first.")

        metrics = {
            'node_count': self.graph.number_of_nodes(),
            'edge_count': self.graph.number_of_edges(),
            'density': nx.density(self.graph),
        }

        # Calculate centrality measures
        if self.graph.number_of_nodes() > 0:
            try:
                metrics['degree_centrality'] = nx.degree_centrality(self.graph)
                metrics['in_degree_centrality'] = nx.in_degree_centrality(self.graph)
                metrics['out_degree_centrality'] = nx.out_degree_centrality(self.graph)
            except Exception:
                # Some graphs might not support certain centrality measures
                pass

        return metrics

File: src/lib/services/graph-builder.ts
/**
 * Graph Builder Service - Constructs temporal communication graphs
 *
 * Takes parsed messages and builds a DiGraph with:
 * - Nodes representing agents
 * - Edges with temporal interaction history
 * - Support for time-based filtering
 */

import { DiGraph } from '@/lib/graph/digraph';
import {
  IntentLabel,
  MessageType,
  type Message,
  type Interaction,
  type EdgeData,
  type NodeData,
  type GraphSnapshot,
  type NodeMetadata,
  type EdgeMetadata,
} from '@/lib/models/types';

export class GraphBuilder {
  private graph: DiGraph<NodeMetadata, EdgeMetadata> | null = null;
  private messages: Message[] = [];

  /**
   * Build a directed graph from a list of messages.
   */
  buildGraph(messages: Message[]): DiGraph<NodeMetadata, EdgeMetadata> {
    if (messages.length === 0) {
      throw new Error('Cannot build graph from empty message list');
    }

    // Sort messages by step_index
    this.messages = [...messages].sort((a, b) => a.step_index - b.step_index);
    this.graph = new DiGraph<NodeMetadata, EdgeMetadata>();

    // Track node metadata separately during construction
    const nodeMetadata: Map<string, NodeMetadata> = new Map();

    // Build nodes and edges
    for (const message of this.messages) {
      // Add sender node
      this.addOrUpdateNode(message.sender, message, nodeMetadata);

      // Add edge if there's a receiver
      if (message.receiver) {
        // Add receiver node
        this.addOrUpdateNode(message.receiver, message, nodeMetadata);

        // Add interaction to edge
        this.addInteraction(message);
      }
    }

    // Update node data in graph
    for (const [nodeId, metadata] of nodeMetadata) {
      this.graph.updateNodeData(nodeId, metadata);
    }

    return this.graph;
  }

  /**
   * Add a new node or update existing node metadata.
   */
  private addOrUpdateNode(
    nodeId: string,
    message: Message,
    nodeMetadata: Map<string, NodeMetadata>
  ): void {
    if (!nodeMetadata.has(nodeId)) {
      nodeMetadata.set(nodeId, {
        id: nodeId,
        label: nodeId,
        message_count: 0,
        messages_sent: 0,
        messages_received: 0,
        first_appearance: message.timestamp,
        last_activity: message.timestamp,
        metadata: {},
      });
      this.graph!.addNode(nodeId);
    }

    const meta = nodeMetadata.get(nodeId)!;

    // Update metadata for sender
    if (message.sender === nodeId) {
      meta.message_count += 1;
      meta.messages_sent += 1;
      if (message.timestamp > meta.last_activity) {
        meta.last_activity = message.timestamp;
      }
    }

    // Track received messages
    if (message.receiver === nodeId) {
      meta.messages_received += 1;
    }
  }

  /**
   * Add an interaction to the edge between sender and receiver.
   */
  private addInteraction(message: Message): void {
    const sender = message.sender;
    const receiver = message.receiver;

    if (!receiver) return;

    // Create interaction object
    const interaction: Interaction = {
      step_index: message.step_index,
      timestamp: message.timestamp,
      intent: this.inferIntent(message),
      message_id: message.step_index,
      metadata: {
        message_type: message.message_type,
        content: message.content || '', // Store full content
        content_preview: message.content?.slice(0, 100) || '', // Keep preview for compatibility
      },
    };

    // Add or update edge
    if (this.graph!.hasEdge(sender, receiver)) {
      // Edge exists - append interaction
      const edgeData = this.graph!.getEdgeData(sender, receiver)!;
      edgeData.interactions.push(interaction);
      edgeData.weight += 1;
    } else {
      // New edge
      this.graph!.addEdge(sender, receiver, {
        source: sender,
        target: receiver,
        interactions: [interaction],
        weight: 1,
      });
    }
  }

  /**
   * Infer the intent label from a message.
   */
  private inferIntent(message: Message): IntentLabel {
    const messageType = message.message_type;

    // Map message types to intent labels
    const typeToIntent: Record<string, IntentLabel> = {
      [MessageType.DELEGATION]: IntentLabel.DELEGATION,
      [MessageType.ACTION]: IntentLabel.COORDINATION,
      [MessageType.RESPONSE]: IntentLabel.INFORMATION_RESPONSE,
      [MessageType.OBSERVATION]: IntentLabel.INFORMATION_RESPONSE,
      [MessageType.THOUGHT]: IntentLabel.UNKNOWN,
      [MessageType.SYSTEM]: IntentLabel.UNKNOWN,
    };

    return typeToIntent[messageType] || IntentLabel.UNKNOWN;
  }

  /**
   * Return a filtered graph containing only interactions up to maxStep.
   * Critical for time-slider functionality.
   */
  filterGraphByStep(maxStep: number): DiGraph<NodeMetadata, EdgeMetadata> {
    if (!this.graph) {
      throw new Error('No graph built yet. Call buildGraph() first.');
    }

    const filtered = this.graph.copy();

    // Filter edges
    const edgesToRemove: Array<{ source: string; target: string }> = [];

    for (const { source, target, data } of filtered.edges()) {
      const filteredInteractions = data.interactions.filter(
        (i) => i.step_index <= maxStep
      );

      if (filteredInteractions.length > 0) {
        // Update edge with filtered interactions
        filtered.setEdgeData(source, target, {
          ...data,
          interactions: filteredInteractions,
          weight: filteredInteractions.length,
        });
      } else {
        edgesToRemove.push({ source, target });
      }
    }

    // Remove edges with no interactions
    for (const { source, target } of edgesToRemove) {
      filtered.removeEdge(source, target);
    }

    // Remove isolated nodes
    const isolatedNodes = filtered.isolates();
    filtered.removeNodesFrom(isolatedNodes);

    return filtered;
  }

  /**
   * Convert the graph to a GraphSnapshot for API response.
   */
  toGraphSnapshot(maxStep?: number): GraphSnapshot {
    if (!this.graph) {
      throw new Error('No graph built yet. Call buildGraph() first.');
    }

    // Use filtered graph if maxStep provided
    const graphToUse =
      maxStep !== undefined ? this.filterGraphByStep(maxStep) : this.graph;

    // Calculate time-based message counts if maxStep is provided
    const timeBasedSentCounts: Record<string, number> = {};
    const timeBasedReceivedCounts: Record<string, number> = {};

    if (maxStep !== undefined) {
      // Count messages sent and received by each agent up to maxStep
      for (const message of this.messages) {
        if (message.step_index <= maxStep) {
          const sender = message.sender;
          timeBasedSentCounts[sender] =
            (timeBasedSentCounts[sender] || 0) + 1;

          if (message.receiver) {
            const receiver = message.receiver;
            timeBasedReceivedCounts[receiver] =
              (timeBasedReceivedCounts[receiver] || 0) + 1;
          }
        }
      }
    }

    // Convert nodes
    const nodes: NodeData[] = [];
    for (const nodeId of graphToUse.nodeIds()) {
      const nodeData = graphToUse.getNodeData(nodeId);
      if (!nodeData) continue;

      // Override counts with time-based counts if maxStep is provided
      const sentCount =
        maxStep !== undefined
          ? timeBasedSentCounts[nodeId] || 0
          : nodeData.messages_sent;
      const receivedCount =
        maxStep !== undefined
          ? timeBasedReceivedCounts[nodeId] || 0
          : nodeData.messages_received;

      nodes.push({
        id: nodeData.id,
        label: nodeData.label,
        message_count: sentCount, // For backwards compatibility
        messages_sent: sentCount,
        messages_received: receivedCount,
        first_appearance: nodeData.first_appearance,
        last_activity: nodeData.last_activity,
        metadata: nodeData.metadata,
      });
    }

    // Convert edges
    const edges: EdgeData[] = graphToUse.edges().map(({ data }) => ({
      source: data.source,
      target: data.target,
      interactions: data.interactions,
      weight: data.weight,
    }));

    const totalSteps =
      this.messages.length > 0
        ? Math.max(...this.messages.map((m) => m.step_index))
        : 0;

    return {
      nodes,
      edges,
      current_step: maxStep ?? null,
      total_steps: totalSteps,
      metadata: {
        node_count: nodes.length,
        edge_count: edges.length,
        message_count: this.messages.length,
      },
    };
  }

  /**
   * Calculate basic graph metrics.
   */
  getGraphMetrics(): Record<string, unknown> {
    if (!this.graph) {
      throw new Error('No graph built yet. Call buildGraph() first.');
    }

    const metrics: Record<string, unknown> = {
      node_count: this.graph.numberOfNodes(),
      edge_count: this.graph.numberOfEdges(),
      density: this.graph.density(),
    };

    // Calculate centrality measures
    if (this.graph.numberOfNodes() > 0) {
      metrics.degree_centrality = this.graph.degreeCentrality();
      metrics.in_degree_centrality = this.graph.inDegreeCentrality();
      metrics.out_degree_centrality = this.graph.outDegreeCentrality();
    }

    return metrics;
  }

  /**
   * Get the messages array.
   */
  getMessages(): Message[] {
    return this.messages;
  }

  /**
   * Get the raw graph.
   */
  getGraph(): DiGraph<NodeMetadata, EdgeMetadata> | null {
    return this.graph;
  }
}
